File: android/core/src/main/kotlin/dev/skyhop/core/HighScoreStore.kt
package dev.skyhop.core

/** Persistence abstraction so the engine and tests stay free of Android APIs. */
interface HighScoreStore {
    suspend fun best(): Int
    suspend fun save(score: Int)
}

class InMemoryHighScoreStore(private var value: Int = 0) : HighScoreStore {
    override suspend fun best(): Int = value
    override suspend fun save(score: Int) {
        if (score > value) value = score
    }
}

File: ios/SkyhopCore/Sources/SkyhopCore/HighScoreStore.swift
import Foundation

/// Persistence abstraction so the view model can be tested without touching disk.
public protocol HighScoreStore: AnyObject {
    func best() -> Int
    func save(_ score: Int)
}

public final class UserDefaultsHighScoreStore: HighScoreStore {
    private let defaults: UserDefaults
    private let key: String

    public init(defaults: UserDefaults = .standard, key: String = "best_score") {
        self.defaults = defaults
        self.key = key
    }

    public func best() -> Int { defaults.integer(forKey: key) }

    public func save(_ score: Int) {
        if score > best() { defaults.set(score, forKey: key) }
    }
}

public final class InMemoryHighScoreStore: HighScoreStore {
    private var value: Int
    public init(_ value: Int = 0) { self.value = value }
    public func best() -> Int { value }
    public func save(_ score: Int) { value = max(value, score) }
}
